pkg/composition/controller2.go

```diff
@@ -0,0 +1,624 @@
+			if info.Namespace != c.namespace && c.namespace != "" {
+				formLogger.Sugar().Infof("Namespace %q doesn't match namespace %q observed by the controller. Skipping formationObjectDef %v.",
+					info.Namespace, c.namespace, info.Location.Label)
+				return true
+			}
+			return false
+		},
+	}
+
+	formationObjectDefList, err := ProcessSD(sd, filteredTransformer)
+
+	foResults := make([]formationObjectResult, 0, len(formationObjectDefList))
+
+	if err != nil {
+		conflict, processRetryable, processErr := c.handleProcessResult(logger, serviceName, sd, foResults, false, false, err)
+		if conflict {
+			return false, nil
+		}
+
+		return processRetryable, processErr
+	}
+
+	var (
+		results        []*formationObjectResult
+		retryable      = false
+		deleteFinished = true
+		processFailure error
+	)
+
+	for _, result := range foResults {
+		formLogger := logger.With(logz.NamespaceName(result.namespace.Namespace))
+		formLogger.Debug("Processing LocationDescriptor for ServiceDescriptor")
+
+		var (
+			newResult        *formationObjectResult
+			processConflict  bool
+			processRetryable bool
+			processErr       error
+		)
+		if sd.ObjectMeta.DeletionTimestamp != nil {
+			var itemFinished bool
+			itemFinished, processConflict, processRetryable, newResult, processErr = c.delete(logger, sd, &result)
+			deleteFinished = deleteFinished && itemFinished
+		} else {
+			processConflict, processRetryable, newResult, processErr = c.processNormalFormationObjectDef2(formLogger, sd, &result)
+		}
+
+		if processConflict {
+			// Something ran into a conflict during processing
+			// The entire ServiceDescriptor will be processed anyway so we just exit
+			return false, nil
+		}
+
+		if processErr != nil {
+			// stop the presses
+			deleteFinished = false
+			processFailure = processErr
+			retryable = processRetryable
+			break
+		}
+		results = append(results, newResult)
+	}
+
+	// If deletion is complete, remove finalizer
+	if sd.ObjectMeta.DeletionTimestamp != nil && deleteFinished {
+		sd.Finalizers = removeServiceDescriptorFinalizer(sd.GetFinalizers())
+		deletionDuration := c.clock.Now().Sub(sd.ObjectMeta.DeletionTimestamp.Time)
+		logger.Info("ServiceDescriptor deleted", zap.Duration("deletionDuration", deletionDuration))
+	}
+
+	nsList, err := c.nsIndexer.ByIndex(nsServiceNameIndex, serviceName)
+	if err != nil {
+		return false, errors.WithStack(err)
+	}
+	var existingNamespaces []*core_v1.Namespace
+	for _, item := range nsList {
+		existingNamespaces = append(existingNamespaces, item.(*core_v1.Namespace))
+	}
+
+	existingLD := make(map[*core_v1.Namespace][]*form_v1.LocationDescriptor)
+	for _, ns := range existingNamespaces {
+		existingLds, err := c.ldIndexer.ByIndex(cache.NamespaceIndex, ns.Name)
+		if err != nil {
+			return false, errors.WithStack(err)
+		}
+
+		for _, obj := range existingLds {
+			existingLD[ns] = append(existingLD[ns], obj.(*form_v1.LocationDescriptor))
+		}
+	}
+
+	st := statusUpdater{
```

Comment:
Define a smaller interface for figuring out the status update. This is more approachable as an interface to use for dry run/validation. This would allow status updates from formation or autowiring to "bubble" back up in a validator without needing the kube api.